In [1]:
import pandas as pd
import datetime as dt
import csv

In [2]:
f = open("guns.csv", "r")

In [3]:
file = pd.read_csv(f)

In [4]:
file.head()

Unnamed: 0  year  month   intent  police sex   age                    race  \
0           1  2012      1  Suicide       0   M  34.0  Asian/Pacific Islander   
1           2  2012      1  Suicide       0   F  21.0                   White   
2           3  2012      1  Suicide       0   M  60.0                   White   
3           4  2012      2  Suicide       0   M  64.0                   White   
4           5  2012      2  Suicide       0   M  31.0                   White   

   hispanic            place  education  
0       100             Home        4.0  
1       100           Street        3.0  
2       100  Other specified        4.0  
3       100             Home        4.0  
4       100  Other specified        2.0

As you can see above, the first row of the data is a header row, which tells you what kind of data is in each column of the CSV file. Each row contains information about the fatality, and the victim. Here's an explanation of each column:

In [5]:
# this python function converts the data frame to a list of lists and
# eliminates the header row
data = file.values.tolist()

In [6]:
data[:5]

[[1,
  2012,
  1,
  'Suicide',
  0,
  'M',
  34.0,
  'Asian/Pacific Islander',
  100,
  'Home',
  4.0],
 [2, 2012, 1, 'Suicide', 0, 'F', 21.0, 'White', 100, 'Street', 3.0],
 [3, 2012, 1, 'Suicide', 0, 'M', 60.0, 'White', 100, 'Other specified', 4.0],
 [4, 2012, 2, 'Suicide', 0, 'M', 64.0, 'White', 100, 'Home', 4.0],
 [5, 2012, 2, 'Suicide', 0, 'M', 31.0, 'White', 100, 'Other specified', 2.0]]

The year column contains information on the year in which gun deaths occurred. We can use this column to calculate how many gun deaths happened in each year.

We can perform this operation by creating a dictionary, then keeping count in the dictionary of how many times each element occurs in the year column.

In [7]:
# use list comprehension to extract the year column from data
years = [row[1] for row in data]

In [8]:
# create an empty dictionary called year_counts
year_counts = {}

In [9]:
#Loop through each element in years.
#If the element isn't a key in year_counts, create it, and set the value to 1.
# If the element is a key in year_counts, increment the value by one.

for year in years:
    if year in year_counts:
        year_counts[year] = year_counts[year] + 1
    else:
        year_counts[year] = 1        

In [10]:
year_counts

{2012: 33563, 2013: 33636, 2014: 33599}

It looks like gun deaths didn't change much by year from 2012 to 2014. Let's see if gun deaths in the US change by month and year. In order to do this, we'll have to create a datetime.datetime object using the year and month columns. 

In [11]:
# Use a list comprehension to create a datetime.datetime object for each row. 
# Assign the result to dates.
dates = [dt.datetime(row[1],row[2], day=1) for row in data]

In [12]:
dates[:5]

[datetime.datetime(2012, 1, 1, 0, 0),
 datetime.datetime(2012, 1, 1, 0, 0),
 datetime.datetime(2012, 1, 1, 0, 0),
 datetime.datetime(2012, 2, 1, 0, 0),
 datetime.datetime(2012, 2, 1, 0, 0)]

In [13]:
# Count up how many times each unique date occurs in dates
date_counts = {}

for date in dates:
    if date in date_counts:
        date_counts[date] = date_counts[date] + 1
    else:
        date_counts[date] = 1

In [14]:
date_counts

{datetime.datetime(2012, 1, 1, 0, 0): 2758,
 datetime.datetime(2012, 2, 1, 0, 0): 2357,
 datetime.datetime(2012, 3, 1, 0, 0): 2743,
 datetime.datetime(2012, 4, 1, 0, 0): 2795,
 datetime.datetime(2012, 5, 1, 0, 0): 2999,
 datetime.datetime(2012, 6, 1, 0, 0): 2826,
 datetime.datetime(2012, 7, 1, 0, 0): 3026,
 datetime.datetime(2012, 8, 1, 0, 0): 2954,
 datetime.datetime(2012, 9, 1, 0, 0): 2852,
 datetime.datetime(2012, 10, 1, 0, 0): 2733,
 datetime.datetime(2012, 11, 1, 0, 0): 2729,
 datetime.datetime(2012, 12, 1, 0, 0): 2791,
 datetime.datetime(2013, 1, 1, 0, 0): 2864,
 datetime.datetime(2013, 2, 1, 0, 0): 2375,
 datetime.datetime(2013, 3, 1, 0, 0): 2862,
 datetime.datetime(2013, 4, 1, 0, 0): 2798,
 datetime.datetime(2013, 5, 1, 0, 0): 2806,
 datetime.datetime(2013, 6, 1, 0, 0): 2920,
 datetime.datetime(2013, 7, 1, 0, 0): 3079,
 datetime.datetime(2013, 8, 1, 0, 0): 2859,
 datetime.datetime(2013, 9, 1, 0, 0): 2742,
 datetime.datetime(2013, 10, 1, 0, 0): 2808,
 datetime.datetime(2013, 11,

The sex and race columns contain potentially interesting information on how gun deaths in the US vary by gender and race. Exploring both of these columns can be done with a similar dictionary counting technique to what we did earlier.

In [15]:
# Count up how many times each item in the sex column occurs.
sex_counts = {}
sex = [row[5] for row in data]

for item in sex:
    if item in sex_counts:
        sex_counts[item] = sex_counts[item] + 1
    else:
        sex_counts[item] = 1


In [16]:
sex_counts

{'F': 14449, 'M': 86349}

In [17]:
# Count up how many times each item in the race column occurs.

race_counts = {}
race = [row[7] for row in data]

for item in race:
    if item in race_counts:
        race_counts[item] = race_counts[item] + 1
    else:
        race_counts[item] = 1

In [18]:
race_counts

{'Asian/Pacific Islander': 1326,
 'Black': 23296,
 'Hispanic': 9022,
 'Native American/Native Alaskan': 917,
 'White': 66237}

In [19]:
# Count up how many times each item in the intent column occurs.
intent_counts = {}
intent = [row[3] for row in data]

for i in intent:
    if i in intent_counts:
        intent_counts[i] = intent_counts[i] + 1
    else:
        intent_counts[i] = 1

intent_counts

{'Suicide': 63175,
 nan: 1,
 'Homicide': 35176,
 'Undetermined': 807,
 'Accidental': 1639}

It appears that the number of gun deaths doesnt vary statistically too very much from month too month. However, we can see that whilte males are the most numerous victims of gun deaths and that suicide is the leading intent.

However, our analysis only gives us the total number of gun deaths by race in the US. Unless we know the proportion of each race in the US, we won't be able to meaningfully compare those numbers. What we really want to get is a rate of gun deaths per 100000 people of each race. In order to do this, we'll need to read in data about what percentage of the US population falls into each racial category. Luckily, we can import some census data to help us out.

In [20]:
# Read in census.csv, and convert to a list of lists. 
# Assign the result to the census variable.

with open("census.csv", "r") as file:
    census = csv.reader(file)
    for row in census:
        print(', '.join(row))

Id, Year, Id, Sex, Id, Hispanic Origin, Id, Id2, Geography, Total, Race Alone - White, Race Alone - Hispanic, Race Alone - Black or African American, Race Alone - American Indian and Alaska Native, Race Alone - Asian, Race Alone - Native Hawaiian and Other Pacific Islander, Two or More Races
cen42010, April 1, 2010 Census, totsex, Both Sexes, tothisp, Total, 0100000US, , United States, 308745538, 197318956, 44618105, 40250635, 3739506, 15159516, 674625, 6984195


In order to get from the raw counts of gun deaths by race to a rate of gun deaths per 100000 people in each race, we'll need to divide the total number of gun deaths by the population of each race. From the census dataset, we know that the number of people in the White racial category is 197318956. We'd divide 66237 by 197318956:

If you do this computation, you'll see that the rate is a very small number, 0.0003356849303419181. It's for this reason that it's typical to express crime statistics as the "rate per 100000". This tells you the number of people in a given group out of every 100000 that were killed by guns in the US. To get this, we just multiply by 100000:

In [24]:
# Manually create a dictionary, mapping that maps each key 
# from race_counts to the population count of the race from census.

mapping = {"Asian/Pacific Islander": 15834141, "Black": 40250635, "Hispanic": 44618105, "Native American/Native Alaskan": 3739506, "White": 197318956}

# Create an empty dictionary, race_per_hundredk
race_per_hundredk = {}

#Loop through each key in race_counts.
#Divide the value associated with the key in race_counts by the value associated with the key in mapping.
#Multiply by 100000.
#Assign the result to the same key in race_per_hundredk.

race_per_hundredk = {k: race_counts[k]/mapping[k] * 100000 for k in race_counts.keys() & mapping}

print(race_per_hundredk)        
        
    



{'Hispanic': 20.220491210910907, 'Native American/Native Alaskan': 24.521955573811088, 'Black': 57.8773477735196, 'Asian/Pacific Islander': 8.374309664161762, 'White': 33.56849303419181}


# Filtering By Intent

We can filter our results, and restrict them to the Homicide intent. This will tell us what the gun-related murder rate per 100000 people in each racial category is. In order to do this, we'll need to redo our work in generating race_counts, but only count rows where the intent was Homicide.

We can do this by first extracting the intent column, then using the enumerate() function to loop through each index and value in the race column. If the value in the same position in intents is Homicide, we'll count the value in the race column.

Finally, we'll use the mapping dictionary to convert from raw counts to rates.

In [30]:
#Extract the intent and race columns using a list comprehension. 

intents = [row[3] for row in data]
races = [row[7] for row in data]

homicide_race_counts = {}


# Use the enumerate() function to loop through each item in races. 
#The position should be assigned to the loop variable i, 
#and the value to the loop variable race.
for i, race in enumerate(races):
    if race not in homicide_race_counts:
        homicide_race_counts[race] = 0
    if intents[i] == "Homicide":
        homicide_race_counts[race] += 1
        
race_per_hundredk = {}
for k,v in homicide_race_counts.items():
    race_per_hundredk[k] = (v / mapping[k]) * 100000

race_per_hundredk


{'Asian/Pacific Islander': 3.530346230970155,
 'Black': 48.471284987180944,
 'Hispanic': 12.627161104219914,
 'Native American/Native Alaskan': 8.717729026240365,
 'White': 4.6356417981453335}

# Findings

It appears that gun related homicides in the US disproportionately affect people in the Black and Hispanic racial categories.

Some areas to investigate further:

* The link between month and homicide rate.
* Homicide rate by gender.
* The rates of other intents by gender and race.
* Gun death rates by location and education.